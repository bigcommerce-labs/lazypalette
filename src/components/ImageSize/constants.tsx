export const enum Messages {
    Dimension = 'Specify a maximum image size for desktop display. We\'ll automatically ' +
        'scale down the image for smaller devices.',
    Lock = 'Constrain Proportions',
}

export const ImageDimensionPattern = (/^[\d]+x[\d]+$/);
