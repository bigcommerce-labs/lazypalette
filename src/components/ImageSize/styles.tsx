import styled from 'styled-components';

import { colors } from '../../styleConstants';

export const ImageSizeModal = styled.div`
  width: 14.0625rem;
  padding: .625rem;
`;

export const SizeModal = styled.div`
  display: flex;
  justify-content: space-between;
  margin-top: .75rem;
`;

export const Axis = styled.div`
  width: 40%;
  position: relative;

  &:first-child:after {
    content: 'x';
    color: ${colors.primary};
    position: absolute;
    top: 2.3125rem;
    left: 6.625rem;
  }
`;
