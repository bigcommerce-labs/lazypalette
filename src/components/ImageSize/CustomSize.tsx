import React, { ChangeEvent, PureComponent } from 'react';

import InputField from '../InputField/InputField';

import { Axis, SizeModal } from './styles';

interface CustomSizeProps {
    defaultValue: string;
    name: string;
    onChange?(configChange: {[key: string]: string}): void;
}

interface CustomSizeState {
    [key: string]: string;
}

class CustomSize extends PureComponent<CustomSizeProps, CustomSizeState> {
    readonly state: CustomSizeState = {
        height: '0',
        width: '0',
    };

    componentDidMount() {
        const { defaultValue } = this.props;
        const [width, height] = defaultValue ? this.props.defaultValue.split('x') : ['0', '0'];
        this.setState({height, width});
    }

    onChange = (event: ChangeEvent<HTMLInputElement>, dimension: string) => {
        const { name } = this.props;
        const { height, width } = this.state;
        const value = event.target.value;

        const dimensions = dimension === 'height' ? `${width}x${value}` : `${value}x${height}`;
        this.setState({ [dimension]: value });
        this.props.onChange!({ [name]: `${dimensions}` });
    };

    render() {
        const { defaultValue } = this.props;
        const dimensions = defaultValue ? defaultValue.split('x') : [0, 0];

        return (
            <SizeModal>
                {['width', 'height'].map((axis, i) => (
                    <Axis key={i}>
                        <InputField
                            defaultValue={+dimensions[i]}
                            label={`Max ${axis}`}
                            required={true}
                            type="number"
                            onChange={e => this.onChange(e, axis)}
                        />
                    </Axis>
                ))}
            </SizeModal>
        );
    }
}

export default CustomSize;
