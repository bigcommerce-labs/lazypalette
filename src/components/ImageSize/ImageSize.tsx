import { SelectBox } from 'pattern-lab';
import React, { ChangeEvent, ChangeEventHandler, Component } from 'react';

import { ImageSizeModal } from './styles';
import CustomSize from './CustomSize';

interface ImageSizeProps {
    label: string;
    options: Array<{
        label: string;
        value: string;
    }>;
    selected: string;
    testId?: string;
    onChange?: ChangeEventHandler<HTMLSelectElement | HTMLInputElement>;
}

interface ImageSizeState {
    defaultValue: string;
    selectedValue: string;
}

class ImageSize extends Component<ImageSizeProps, ImageSizeState> {
    readonly state: ImageSizeState = {
        defaultValue: '1x1',
        selectedValue: 'optimized',
    };

    componentDidMount() {
        if (this.props.selected) {
            this.setState({ defaultValue: this.props.selected });
        }
    }

    componentDidUpdate(prevProps: ImageSizeProps) {
        if (prevProps.selected !== this.props.selected) {
            this.setState({ defaultValue: this.props.selected });
        }
    }

    handleSelectChange = (event: ChangeEvent<HTMLSelectElement>) => {
        const value = event.target.value;
        const { defaultValue } = this.state;
        this.setState({ selectedValue: value });

        const selectEvent = value === 'custom' ? {...event, target: {...event.target, value: defaultValue}} : event;

        if (this.props.onChange) {
            this.props.onChange(selectEvent);
        }
    };

    handleCustomChange = (event: ChangeEvent<HTMLInputElement>) => {
        if (this.state.selectedValue === 'optimized') {
            this.setState({ selectedValue: 'custom' });
        }
        this.props.onChange!(event);
    };

    render() {
        const { label, options, selected, testId } = this.props;
        const { selectedValue } = this.state;
        const optionExists = options.some(({value}) => value === selected);
        const optionSelected = (optionExists && selectedValue !== 'custom') ? selected : 'custom';

        return (
            <ImageSizeModal>
                <SelectBox
                    label={label}
                    options={options}
                    selected={optionSelected}
                    onChange={this.handleSelectChange}
                    testId={testId}
                />
                {optionSelected === 'custom' &&
                    <CustomSize defaultValue={selected} onChange={this.handleCustomChange} />
                }
            </ImageSizeModal>
        );
    }

}

export default ImageSize;
