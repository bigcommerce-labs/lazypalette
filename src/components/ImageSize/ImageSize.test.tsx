import { shallow } from 'enzyme';
import React from 'react';
import ImageSize from './ImageSize';

it('renders', () => {
  const testItems = [
    {
      label: 'Optimized for theme',
      value: '100x100',
    },
    {
      label: 'Specify dimensions',
      value: 'foo',
    },
  ];
  const imageSize = shallow(
    <ImageSize
        label="test"
        options={testItems}
        selected="foo"
        name="test"
    />
  );

  expect(imageSize).toMatchSnapshot();
});
