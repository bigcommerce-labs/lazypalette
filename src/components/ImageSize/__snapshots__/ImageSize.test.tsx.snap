// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageSize /> test should render <SelectBox /> 1`] = `
<styled.div>
  <SelectBox
    label="test"
    name="test"
    onChange={[Function]}
    options={
      Array [
        Object {
          "label": "Original (as uploaded)",
          "value": "original",
        },
        Object {
          "label": "Optimized for theme",
          "value": "250x100",
        },
        Object {
          "label": "Specify dimensions",
          "value": "custom",
        },
      ]
    }
    selected="custom"
  />
  <CustomSize
    defaultValue="foo"
    name="test"
    onChange={[Function]}
  />
</styled.div>
`;

exports[`<ImageSize /> test should render <SelectBox /> and <CustomSize /> components 1`] = `
<styled.div>
  <SelectBox
    label="test"
    name="test"
    onChange={[Function]}
    options={
      Array [
        Object {
          "label": "Original (as uploaded)",
          "value": "original",
        },
        Object {
          "label": "Optimized for theme",
          "value": "250x100",
        },
        Object {
          "label": "Specify dimensions",
          "value": "custom",
        },
      ]
    }
    selected="custom"
  />
  <CustomSize
    defaultValue="custom"
    name="test"
    onChange={[Function]}
  />
</styled.div>
`;
